### ISSUE TITLE
Webhook payload duplicate properties

### ISSUE THREAD

**Issue by iiinnniii** [go-gitea/gitea#22072]
### Description

These properties do exist twice for some reason within the `pusher` object within the payload of the Gitea Webhook.

```
last_login: string;
created: string;
restricted: boolean;
active: boolean;
prohibit_login: boolean;
location: string;
website: string;
description: string;
visibility: string;
```

Webhook config:

Target URL: http://149.56.10.3:3000/build
HTTP Method: POST
POST Content Type: application/json
Secret:
Trigger On: Push Events
Branch Filter: main
Active: Yes

I cannot reproduce on https://try.gitea.io/, because I do not have a domain für my build server, but just an ip which is not exposed to the public.

Thanks.

### Gitea Version

Gitea version 1.17.3 built with GNU Make 4.1, go1.18.5 : bindata, sqlite, sqlite_unlock_notify, pam, cert

### Can you reproduce the bug on the Gitea demo site?

No

### Screenshots

[Link to screenshot](https://ibb.co/0DJqPNr)

### Operating System

Ubuntu Server 22.04.01 LTS

### How are you running Gitea?

[Gitea snap package](https://snapcraft.io/gitea)

### Database

SQLite

**Comment by jolheiser** [comment]
Do you have an example payload? 
Is it a generic gitea webhook?
Are you sure you aren't seeing `pusher` and `sender` objects that are near each other?

**Comment by gernotpokorny** [comment]
> Is it a generic gitea webhook?

The Webhook got defined like this with the data mentioned in the OP:

Repo > Webhooks > Add Webhook > Gitea

> Do you have an example payload?

Example payload:

```
{
  ref: 'refs/heads/main',
  before: 'f8d1d37f6ab7c4d02fc0dce5cd90e6f893f93fc8',
  after: '900e52ccd1281b9cf9dd55b3f8327e851e682852',
  compare_url: 'http://149.56.10.2:3000/gernot/myproject/compare/f8d1d37f6ab7c4d02fc0dce5cd90e6f893f93fc8...900e52ccd1281b9cf9dd55b3f8327e851e682852',
  commits: [
    {
      id: '900e52ccd1281b9cf9dd55b3f8327e851e682852',
      message: 'Removed empty line\n',
      url: 'http://149.56.10.2:3000/gernot/myproject/commit/900e52ccd1281b9cf9dd55b3f8327e851e682852',
      author: [Object],
      committer: [Object],
      verification: null,
      timestamp: '2022-12-08T18:00:58+01:00',
      added: [],
      removed: [],
      modified: [Array]
    }
  ],
  head_commit: {
    id: '900e52ccd1281b9cf9dd55b3f8327e851e682852',
    message: 'Removed empty line\n',
    url: 'http://149.56.10.2:3000/gernot/myproject/commit/900e52ccd1281b9cf9dd55b3f8327e851e682852',
    author: {
      name: 'Gernot Pokorny',
      email: 'myemail@gmail.com',
      username: 'gernot'
    },
    committer: {
      name: 'Gernot Pokorny',
      email: 'myemail@gmail.com',
      username: 'gernot'
    },
    verification: null,
    timestamp: '2022-12-08T18:00:58+01:00',
    added: [],
    removed: [],
    modified: [ 'src/index.tsx' ]
  },
  repository: {
    id: 7,
    owner: {
      id: 1,
      login: 'gernot',
      full_name: '',
      email: 'myemail@gmail.com',
      avatar_url: 'http://149.56.10.2:3000/avatar/d4a3e1c68b3f25d49060b55911c7afcc',
      language: '',
      is_admin: false,
      last_login: '0001-01-01T00:00:00Z',
      created: '2022-08-27T03:10:23Z',
      restricted: false,
      active: false,
      prohibit_login: false,
      location: '',
      website: '',
      description: '',
      visibility: 'public',
      followers_count: 0,
      following_count: 0,
      starred_repos_count: 0,
      username: 'gernot'
    },
    name: 'myproject',
    full_name: 'gernot/myproject',
    description: '',
    empty: false,
    private: true,
    fork: false,
    template: false,
    parent: null,
    mirror: false,
    size: 769,
    language: '',
    languages_url: 'http://149.56.10.2:3000/api/v1/repos/gernot/myproject/languages',
    html_url: 'http://149.56.10.2:3000/gernot/myproject',
    ssh_url: 'root@localhost:gernot/myproject.git',
    clone_url: 'http://149.56.10.2:3000/gernot/myproject.git',
    original_url: 'http://149.56.10.3:3000/gernot/myproject.git',
    website: '',
    stars_count: 0,
    forks_count: 0,
    watchers_count: 1,
    open_issues_count: 9,
    open_pr_counter: 0,
    release_counter: 0,
    default_branch: 'main',
    archived: false,
    created_at: '2022-08-28T21:48:40Z',
    updated_at: '2022-12-08T16:27:18Z',
    permissions: { admin: true, push: true, pull: true },
    has_issues: true,
    internal_tracker: {
      enable_time_tracker: true,
      allow_only_contributors_to_track_time: true,
      enable_issue_dependencies: true
    },
    has_wiki: true,
    has_pull_requests: true,
    has_projects: true,
    ignore_whitespace_conflicts: false,
    allow_merge_commits: true,
    allow_rebase: true,
    allow_rebase_explicit: true,
    allow_squash_merge: true,
    default_merge_style: 'merge',
    avatar_url: '',
    internal: false,
    mirror_interval: '',
    mirror_updated: '0001-01-01T00:00:00Z',
    repo_transfer: null
  },
  pusher: {
    id: 1,
    login: 'gernot',
    full_name: '',
    email: 'myemail@gmail.com',
    avatar_url: 'http://149.56.10.2:3000/avatar/d4a3e1c68b3f25d49060b55911c7afcc',
    language: '',
    is_admin: false,
    last_login: '0001-01-01T00:00:00Z',
    created: '2022-08-27T03:10:23Z',
    restricted: false,
    active: false,
    prohibit_login: false,
    location: '',
    website: '',
    description: '',
    visibility: 'public',
    last_login: '0001-01-01T00:00:00Z',
    created: '2022-08-27T03:10:23Z',
    restricted: false,
    active: false,
    prohibit_login: false,
    location: '',
    website: '',
    description: '',
    visibility: 'public',
    followers_count: 0,
    following_count: 0,
    starred_repos_count: 0,
    username: 'gernot'
  }
}
```

> Are you sure you aren't seeing pusher and sender objects that are near each other?

I think so, I posted the payload.

**Comment by jolheiser** [comment]
I don't see a `sender` object at all, how are you viewing this JSON? Is something for some reason folding the output or truncating it?

**Comment by gernotpokorny** [comment]
My server does use nodejs, `express` and `body-parser` and basically has just one route where I log this data.

**Comment by gernotpokorny** [comment]
src\index.ts:

```
import express from 'express';
import bodyParser from 'body-parser';

// router
import { buildRouter } from './routes/build';

const app = express();
app.use(bodyParser.json());

app.use('/build', buildRouter);

app.listen(3000);
```

src\routes\build.ts:

```
import { Router } from 'express';

interface PostBuildRequest {
    body: any;
}

export const buildRouter = Router();

buildRouter.post('/', (req: PostBuildRequest, res, next) => {
    console.log(req.body);
    res.status(201).json({ message: 'Build successfull.', id: 'dsfdsf43453' });
});
```

**Comment by gernotpokorny** [comment]
Yes, there is a `sender` and `pusher` object and they look correct there.

**Comment by gernotpokorny** [comment]
I guess this can then be closed and it is an issue with `body-parser` or so, which I have to figure out. I know it's out of scope, but do you have an idea what's causing this problem with this code?

**Comment by gernotpokorny** [comment]
Headers:

```
Request URL: http://149.56.10.3:3000/build
Request method: POST
Content-Type: application/json
X-GitHub-Delivery: c3f017a1-f27d-40d2-81a1-025f76fc20cf
X-GitHub-Event: push
X-GitHub-Event-Type: push
X-Gitea-Delivery: c3f017a1-f27d-40d2-81a1-025f76fc20cf
X-Gitea-Event: push
X-Gitea-Event-Type: push
X-Gitea-Signature: 
X-Gogs-Delivery: c3f017a1-f27d-40d2-81a1-025f76fc20cf
X-Gogs-Event: push
X-Gogs-Event-Type: push
X-Gogs-Signature: 
X-Hub-Signature: sha1=
X-Hub-Signature-256: sha256=
```


Content:

```
{
  "ref": "refs/heads/main",
  "before": "f8d1d37f6ab7c4d02fc0dce5cd90e6f893f93fc8",
  "after": "900e52ccd1281b9cf9dd55b3f8327e851e682852",
  "compare_url": "http://149.56.10.2:3000/gernot/myproject/compare/f8d1d37f6ab7c4d02fc0dce5cd90e6f893f93fc8...900e52ccd1281b9cf9dd55b3f8327e851e682852",
  "commits": [
    {
      "id": "900e52ccd1281b9cf9dd55b3f8327e851e682852",
      "message": "Removed empty line\n",
      "url": "http://149.56.10.2:3000/gernot/myproject/commit/900e52ccd1281b9cf9dd55b3f8327e851e682852",
      "author": {
        "name": "Gernot Pokorny",
        "email": "myemail@gmail.com",
        "username": "gernot"
      },
      "committer": {
        "name": "Gernot Pokorny",
        "email": "myemail@gmail.com",
        "username": "gernot"
      },
      "verification": null,
      "timestamp": "2022-12-08T18:00:58+01:00",
      "added": [],
      "removed": [],
      "modified": [
        "src/index.tsx"
      ]
    }
  ],
  "head_commit": {
    "id": "900e52ccd1281b9cf9dd55b3f8327e851e682852",
    "message": "Removed empty line\n",
    "url": "http://149.56.10.2:3000/gernot/myproject/commit/900e52ccd1281b9cf9dd55b3f8327e851e682852",
    "author": {
      "name": "Gernot Pokorny",
      "email": "myemail@gmail.com",
      "username": "gernot"
    },
    "committer": {
      "name": "Gernot Pokorny",
      "email": "myemail@gmail.com",
      "username": "gernot"
    },
    "verification": null,
    "timestamp": "2022-12-08T18:00:58+01:00",
    "added": [],
    "removed": [],
    "modified": [
      "src/index.tsx"
    ]
  },
  "repository": {
    "id": 7,
    "owner": {
      "id": 1,
      "login": "gernot",
      "full_name": "",
      "email": "myemail@gmail.com",
      "avatar_url": "http://149.56.10.2:3000/avatar/d4a3e1c68b3f25d49060b55911c7afcc",
      "language": "",
      "is_admin": false,
      "last_login": "0001-01-01T00:00:00Z",
      "created": "2022-08-27T03:10:23Z",
      "restricted": false,
      "active": false,
      "prohibit_login": false,
      "location": "",
      "website": "",
      "description": "",
      "visibility": "public",
      "followers_count": 0,
      "following_count": 0,
      "starred_repos_count": 0,
      "username": "gernot"
    },
    "name": "myproject",
    "full_name": "gernot/myproject",
    "description": "",
    "empty": false,
    "private": true,
    "fork": false,
    "template": false,
    "parent": null,
    "mirror": false,
    "size": 769,
    "language": "",
    "languages_url": "http://149.56.10.2:3000/api/v1/repos/gernot/myproject/languages",
    "html_url": "http://149.56.10.2:3000/gernot/myproject",
    "ssh_url": "root@localhost:gernot/myproject.git",
    "clone_url": "http://149.56.10.2:3000/gernot/myproject.git",
    "original_url": "http://149.56.10.3:3000/gernot/myproject.git",
    "website": "",
    "stars_count": 0,
    "forks_count": 0,
    "watchers_count": 1,
    "open_issues_count": 9,
    "open_pr_counter": 0,
    "release_counter": 0,
    "default_branch": "main",
    "archived": false,
    "created_at": "2022-08-28T21:48:40Z",
    "updated_at": "2022-12-08T16:27:18Z",
    "permissions": {
      "admin": true,
      "push": true,
      "pull": true
    },
    "has_issues": true,
    "internal_tracker": {
      "enable_time_tracker": true,
      "allow_only_contributors_to_track_time": true,
      "enable_issue_dependencies": true
    },
    "has_wiki": true,
    "has_pull_requests": true,
    "has_projects": true,
    "ignore_whitespace_conflicts": false,
    "allow_merge_commits": true,
    "allow_rebase": true,
    "allow_rebase_explicit": true,
    "allow_squash_merge": true,
    "default_merge_style": "merge",
    "avatar_url": "",
    "internal": false,
    "mirror_interval": "",
    "mirror_updated": "0001-01-01T00:00:00Z",
    "repo_transfer": null
  },
  "pusher": {
    "id": 1,
    "login": "gernot",
    "full_name": "",
    "email": "myemail@gmail.com",
    "avatar_url": "http://149.56.10.2:3000/avatar/d4a3e1c68b3f25d49060b55911c7afcc",
    "language": "",
    "is_admin": false,
    "last_login": "0001-01-01T00:00:00Z",
    "created": "2022-08-27T03:10:23Z",
    "restricted": false,
    "active": false,
    "prohibit_login": false,
    "location": "",
    "website": "",
    "description": "",
    "visibility": "public",
    "followers_count": 0,
    "following_count": 0,
    "starred_repos_count": 0,
    "username": "gernot"
  },
  "sender": {
    "id": 1,
    "login": "gernot",
    "full_name": "",
    "email": "myemail@gmail.com",
    "avatar_url": "http://149.56.10.2:3000/avatar/d4a3e1c68b3f25d49060b55911c7afcc",
    "language": "",
    "is_admin": false,
    "last_login": "0001-01-01T00:00:00Z",
    "created": "2022-08-27T03:10:23Z",
    "restricted": false,
    "active": false,
    "prohibit_login": false,
    "location": "",
    "website": "",
    "description": "",
    "visibility": "public",
    "followers_count": 0,
    "following_count": 0,
    "starred_repos_count": 0,
    "username": "gernot"
  }
}
```

**Comment by jolheiser** [comment]
Hmm, when I `console.log` that directly it seems to come out mostly okay (albeit the linebreaks make it look weird)
So unfortunately this seems to be related to `body-parser`, which I don't have experience with. You could potentially ask over on their [issue tracker](https://github.com/expressjs/body-parser/issues), however.
Sorry I couldn't be more help.

**Comment by wxiaoguang** [comment]
I think you are cheated by your "terminal"

Some terminals can not handle scrolling correctly, you may read incorrect result.

I suggest you to output the content into a file and download it to your local computer to re-check it.

**Comment by wxiaoguang** [comment]
For example, if the middle part disappears in your "terminal", you will see:

```
  "pusher": {
    "id": 1,
    "login": "gernot",
    "full_name": "",
    "email": "myemail@gmail.com",
    "avatar_url": "http://149.56.10.2:3000/avatar/d4a3e1c68b3f25d49060b55911c7afcc",
    "language": "",
    "is_admin": false,
    "last_login": "0001-01-01T00:00:00Z",
    "created": "2022-08-27T03:10:23Z",
    "restricted": false,
    "active": false,
    "prohibit_login": false,
    "location": "",
    "website": "",
    "description": "",
    "visibility": "public",
// some lines disappear from here
    "followers_count": 0,
    "following_count": 0,
    "starred_repos_count": 0,
    "username": "gernot"
  },
  "sender": {
    "id": 1,
    "login": "gernot",
    "full_name": "",
    "email": "myemail@gmail.com",
    "avatar_url": "http://149.56.10.2:3000/avatar/d4a3e1c68b3f25d49060b55911c7afcc",
    "language": "",
    "is_admin": false,
// some lines disappear to here
    "last_login": "0001-01-01T00:00:00Z",
    "created": "2022-08-27T03:10:23Z",
    "restricted": false,
    "active": false,
    "prohibit_login": false,
    "location": "",
    "website": "",
    "description": "",
    "visibility": "public",
    "followers_count": 0,
    "following_count": 0,
    "starred_repos_count": 0,
    "username": "gernot"
  }
```

**Comment by gernotpokorny** [comment]
I think you are right. I started the debugger and set a breakpoint and there I see the `sender` and `pusher` object and they both look correct.
